src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,760 @@
+            assert(counts_size >= layout.cache_line_size);
+            assert(counts_size % layout.cache_line_size == 0);
+
+            const clocks_size = @divExact(sets * clock_hand_bits, 8);
+            assert(clocks_size >= layout.cache_line_size);
+            assert(clocks_size % layout.cache_line_size == 0);
+
+            const tags = try allocator.alloc(Tag, value_count_max);
+            errdefer allocator.free(tags);
+
+            const values = try allocator.allocAdvanced(
+                Value,
+                value_alignment,
+                value_count_max,
+                .exact,
+            );
+            errdefer allocator.free(values);
+
+            const counts = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(counts);
+
+            const clocks = try allocator.alloc(u64, @divExact(clocks_size, @sizeOf(u64)));
+            errdefer allocator.free(clocks);
+
+            var self = Self{
+                .sets = sets,
+                .tags = tags,
+                .values = values,
+                .counts = .{ .words = counts },
+                .clocks = .{ .words = clocks },
+            };
+
+            self.reset();
+
+            return self;
+        }
+
+        pub fn deinit(self: *Self, allocator: mem.Allocator) void {
+            assert(self.sets > 0);
+            self.sets = 0;
+
+            allocator.free(self.tags);
+            allocator.free(self.values);
+            allocator.free(self.counts.words);
+            allocator.free(self.clocks.words);
+        }
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts.words, 0);
+            mem.set(u64, self.clocks.words, 0);
+        }
+
+        pub fn get(self: *Self, key: Key) ?*align(value_alignment) Value {
+            const set = self.associate(key);
+            const way = self.search(set, key) orelse return null;
+
+            const count = self.counts.get(set.offset + way);
+            self.counts.set(set.offset + way, count +| 1);
+
+            return @alignCast(value_alignment, &set.values[way]);
+        }
+
+        /// Remove a key from the set associative cache if present.
+        pub fn remove(self: *Self, key: Key) void {
+            const set = self.associate(key);
+            const way = self.search(set, key) orelse return;
+
+            self.counts.set(set.offset + way, 0);
+        }
+
+        /// If the key is present in the set, returns the way. Otherwise returns null.
+        inline fn search(self: *Self, set: Set, key: Key) ?usize {
+            const ways = search_tags(set.tags, set.tag);
+
+            var it = BitIterator(Ways){ .bits = ways };
+            while (it.next()) |way| {
+                const count = self.counts.get(set.offset + way);
+                if (count > 0 and equal(key_from_value(set.values[way]), key)) {
```

Comment:
It would be good to improve the test coverage for the `equal()` condition here.

I was bashing the code and found that `if (count > 0 and (true or equal(key_from_value(set.values[way]), key))) {` still passed all the tests.

Reply:
I added a second run of the same tests with a "hash" function that always returns 0 in 67bd687137864e1ca830e098c5eec1a91b8e49da. The tests now fail if that change is made.